### DISCUSSION TITLE
How to count the last 5 days in query Clickhouse

### DISCUSSION THREAD

**Discussion by lucasenabit** [ClickHouse/ClickHouse#52896] [Q&A] [answered]
hello everyone, I have a problem to generate a consolidated query. My scenario is as follows, I pass a date range for example
WHERE day >= '2023-06-26' and day <= '2023-07-01' , within that precise range that returns the total records for the last 5 days for each day.

Exemple: 

`select toDate(Created) a, count() from hml.gold_table
 where Created >=  date_add(day,-5,toDate('2023-06-26')) and Created <= toDate('2023-07-01')
group by toDate(Created)
order by a`

Result: 

<img width="303" alt="Screenshot 2023-08-01 at 21 15 19" src="https://github.com/ClickHouse/ClickHouse/assets/108435054/43e4a853-deb6-4a80-868e-96cbb0a80a29">

In this case, for example for the date 2023-06-25, I need you to return the consolidated for the last 5 days, that is, the sum of the total records between the 21st and the 25th. For the 26th, the sum of the total from the 22nd to the 26th, to the 27th from the 23rd to the 27th and so on.

After trying several times, I got close to what I'm wanting, but in the select below, it's adding up all the last days and not just the last 5 of each day.

`
select
dd.1 as date,
arraySum(x -> x.1 <  dd.1 ? x.2: 0 , days) as prev
--arraySum(x -> x.1 >= dd.1 ? x.2: 0, days) as next
from (
    select groupArray((day, s)) days from (
        select toDate(Created) day, count() s from (
            -- table that generates random dates, numbers
            select Created from hml.gold_tabela_de_consulta
            WHERE Created >=  date_add(day,-5,toDate('2023-06-26')) and Created <= '2023-07-01'
            --select now() + number*3600 as ts, number as amount from numbers(1000)
        ) group by day
    )
) array join days as dd
WHERE date >=  '2023-06-26' and date <= '2023-07-01'
order by date asc`

this is my result:

<img width="250" alt="Screenshot 2023-08-01 at 21 23 08" src="https://github.com/ClickHouse/ClickHouse/assets/108435054/9eb1cbce-1506-4103-8aff-9acfe65b7db2">

However, as I said, the result of only the first line is correct, the others add up from the first day to the return day of each date.

The expected result should be like this:

<img width="573" alt="Screenshot 2023-08-01 at 21 26 15" src="https://github.com/ClickHouse/ClickHouse/assets/108435054/981bea96-0aa1-433f-b6ae-b923c59d49a8">

**Comment by UnamedRus** [comment, marked as answer]
Use The Window functions, Luke

```
SELECT
    number,
    sum(number),
    sum(sum(number)) OVER (ORDER BY number ASC ROWS BETWEEN 4 PRECEDING AND CURRENT ROW) AS running_5,
    groupArray(sum(number)) OVER (ORDER BY number ASC ROWS BETWEEN 4 PRECEDING AND CURRENT ROW) AS values
FROM numbers(10)
GROUP BY number

Query id: 39e5eddf-f386-4c2a-a28c-6882edbd6878

┌─number─┬─sum(number)─┬─running_5─┬─values──────┐
│      0 │           0 │         0 │ [0]         │
│      1 │           1 │         1 │ [0,1]       │
│      2 │           2 │         3 │ [0,1,2]     │
│      3 │           3 │         6 │ [0,1,2,3]   │
│      4 │           4 │        10 │ [0,1,2,3,4] │
│      5 │           5 │        15 │ [1,2,3,4,5] │
│      6 │           6 │        20 │ [2,3,4,5,6] │
│      7 │           7 │        25 │ [3,4,5,6,7] │
│      8 │           8 │        30 │ [4,5,6,7,8] │
│      9 │           9 │        35 │ [5,6,7,8,9] │
└────────┴─────────────┴───────────┴─────────────┘
```

**Comment by lucasenabit** [comment]
great, I went a long way to try to do something so simple. Thank you very much.

Taking advantage, in this example the first dates do not bring the previous values. In my example, is there any way to add the last 5 days of the 21st, even using it as the beginning of my filter? For example, the 21, 22, 23, 24 will not return the last 5


`where Created >=  toDate('2023-06-21')) and Created <= toDate('2023-07-01')`

**Reply by UnamedRus** [reply]
> In my example, is there any way to add the last 5 days of the 21st, even using it as the beginning of my filter? 

No, you need to query 5 more previous days.

Later you can remove them with WHERE condition in top level query

**Comment by lucasenabit** [comment]
Yes, I made that adjustment. I used a subquery to add -5 day at the beginning and then filter in the resultset. Thank you for your help !
